Added support for using Portable Dialog

ravensorb opened the pull request:
Inital version PortableDialog to use NuGet Client framework and to use MVVM pattern
Added support for Xamarin in PortableDIalog

ravensorb commented:
This should build -- is it possible that nuget on appveyor cannot download the Nuget Client v3 packages?

304NotModified commented on PackageExplorer/PortableLibraryDialog.xaml.cs:
```diff
@@ -53,9 +39,9 @@ private void CancelButton_Click(object sender, RoutedEventArgs e)
-            var _allCheckBoxes = new CheckBox[] { NetCheckBox, SilverlightCheckBox, WindowsCheckBox, WPCheckBox };
-            var count = _allCheckBoxes.Count(p => p.IsChecked == true);
-            OKButton.IsEnabled = count >= 2;
+            //var _allCheckBoxes = new CheckBox[] { NetCheckBox, SilverlightCheckBox, WindowsCheckBox, WPCheckBox, XamarinAndroid, XamariniOS };
```
please remove commented code.

304NotModified commented:
Thanks @ravensorb !

Too bad we cannot accept this with the current file headers. We don't like the possible copyright issues. Proposal: remove the headers, your are of course the author of the commits.

If that is possible, I will try to fix AppVeyor. Probable needs some config.


PS: compare without newline changes: https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/pull/72/files?w=1

ravensorb commented:
Ok, I found out that the Nuget package I referenced was actually an old one :) So I updated that, removed the headers, removed the commented code, and did some more clean up.  The only thing outstanding is I am not 100% sure yet how to use the NuGet framework library to VALIDATE the selected platforms so I am sticking with a simple option right now.

304NotModified commented:
OK thanks! Will look at the code in-depth

304NotModified commented on PackageExplorer/Models/NuGetFrameworkModel.cs:
```diff
@@ -0,0 +1,255 @@
+										 where platform.IsSelected
+										 select platform.SelectedItem?.Framework ?? platform?.Framework);
+
+				return selectedPlatforms;
```
On other code parts Linq Extensions are used instead of the query syntax. I also prefer that. Can you rewrite that to 

`PortableFrameworks.Where(...).Select(..)` ? (Reshaper can do that for your)

ravensorb replied:
Done

304NotModified commented on PackageExplorer/Models/NuGetFrameworkModel.cs:
```diff
@@ -0,0 +1,255 @@
+		/// </summary>
+		private void InitializePCLModel()
+		{
+			var frameWork = new NuGetTargetFrameworkItem { Name = ".NET Framework" };
```
There a lot of code duplication here. Can we remove that? I was thinking of a dictionary which contains (at least) tne name & shortname

ravensorb replied:
Not sure I follow -- the entire point of this set of extensions is to leverage the existing objects within the NuGet class libraries.  These extensions just make it a lot easier to work with for the kinds of operations needed for the UI.

304NotModified replied:
```
frameWork.Items.Add(new NuGetTargetFrameworkItem { Name = ".NET Framework 4 and higher", ShortName = "net4", Framework = Frameworks.FindByIdentifierShortName("net40").FirstOrDefault() });
```

is duplicate of
```
frameWork.Items.Add(new NuGetTargetFrameworkItem { Name = ".NET Framework 4.0.3 and higher", ShortName = "net403", Framework = Frameworks.FindByIdentifierShortName("net403").FirstOrDefault() });
```

etc

The non-duplicate things are  ".NET Framework 4.0.3 and higher" (Name) and "net403" (ShortName)

ravensorb replied:
They are actually different when it comes to portable framework identifiers and paths.

304NotModified commented on PackageExplorer/Models/NuGetFrameworkModel.cs:
```diff
@@ -0,0 +1,255 @@
+	/// Class NuGetTargetFrameworkItem.
+	/// </summary>
+	[PropertyChanged.ImplementPropertyChanged]
+	public class NuGetTargetFrameworkItem
```
Please move to separate file

ravensorb replied:
Done

304NotModified commented on PackageExplorer/Models/NuGetFrameworkModel.cs:
```diff
@@ -0,0 +1,255 @@
+		public string Name { get; set; }
+
+		/// <summary>
+		/// Gets or sets the short name.
```
Thanks for all the docs, but this won't help much. What is a short name? (maybe add an example?)

304NotModified commented on PackageExplorer/Models/NuGetFrameworkModel.cs:
```diff
@@ -0,0 +1,255 @@
+		/// <returns>System.String.</returns>
+		public static string AsTargetedPlatformPath(this ObservableCollection<NuGetTargetFrameworkItem> items)
+		{
+			var selectedPlatforms = (from platform in items
```
This is duplication code. Please remove it

ravensorb replied:
Not sure I understand -- there are two extension methods based on which type of collection is being used.  Where is the duplication?

ravensorb commented:
Just updated based on you feedback -- thanks!

304NotModified commented on PackageExplorer/Utilities/NuGetExtensions.cs:
```diff
@@ -0,0 +1,271 @@
+		public static NuGetFramework FindByFrameworkName(this IFrameworkMappings frameworkMapping, string frameworkName)
+		{
+			var result = frameworkMapping.EquivalentFrameworks.FirstOrDefault(x =>
+				string.Compare(x.Key.Framework, frameworkName, StringComparison.InvariantCultureIgnoreCase) == 0 ||
```
Why are we using compare, and not equals?

ravensorb replied:
equals is case and locale sensitive -- Compares is more of a "best practice" from a development perspective.  Since the NuGet packages have the potential to be localized, I didn't want anything to break :)

304NotModified replied:
> equals is case and locale sensitive

That's not the case when passing ` StringComparison.InvariantCultureIgnoreCase` to Equals?

ravensorb replied:
I suppose however it assumes that the object in question is not null - where compare handles null on either operand.  I guess I could switch to something like 
`x.Key.Framework?.Equals(frameworkName, StringComparison.InvariantCulterIgnoreCase)`
Just not sure what the really gets us?

304NotModified commented on PackageExplorer/Utilities/NuGetExtensions.cs:
```diff
@@ -0,0 +1,271 @@
+		/// </summary>
+		/// <param name="frameworks">The frameworks.</param>
+		/// <param name="identifierShortName">Short name of the identifier.</param>
+		/// <returns>IEnumerable&lt;NuGetFramework&gt;.</returns>
```
You can use the `<see` tag instead of `IEnumerable&lt;NuGetFramework&gt;.` here.

But in this Case these `<returns>` won't add anything, so please remove it.

ravensorb replied:
I usually keep my comments in a specific format as it makes generate documentation more consistent.  Also the "see" tag doesn't always work correctly on IEnumerable return results as most of the time it takes you to the IEnumerable documetnation rather than the object that is enumerable.

304NotModified replied:
Well do you agree that it won't add anything now? (for the end-user)

ravensorb replied:
I updated the comment in the latest push -- should have more info in almost all of the returns comments to add more value.  Will that be ok?

304NotModified commented on PackageExplorer/packages.config:
```diff
@@ -0,0 +1,7 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Fody" version="1.29.4" targetFramework="net45" developmentDependency="true" />
```
What's Fody?

ravensorb replied:
[Fody](https://github.com/Fody/Fody) is a tool that makes certain things easier :smile: when it comes to getting right of "boiler plate" code.  I use it to handle the code for [PropertChanged](https://github.com/Fody/PropertyChanged) notifications a lot as the code for that is almsot 100% redundant in all cases.

304NotModified replied:
Cool! will read into it

ravensorb commented:
Everything ok to merge this into the main branch?

304NotModified commented on PackageExplorer/Models/NuGetFrameworkModel.cs:
```diff
@@ -0,0 +1,123 @@
+
+			frameWork = new NuGetTargetFrameworkItem { Name = "Xamarin For Android", ShortName = "MonoAndroid", Framework = Frameworks.FindByIdentifierShortName("MonoAndroid").FirstOrDefault() };
+			PortableFrameworks.Add(frameWork);
+		}
```
Please try to remove the code- / config duplication in this method. My previous comment on this issue has been lost I see.

ravensorb replied:
I am not sure I understand, there is nothing in here that is duplicate.  This is the core of the most recent enhancement, it moves ALL of the target platforms into a single are and leverages the NuGet libraries. Everything in here is needed to support the different platforms within the selection dialog.  Eventually I'd love to move this into the NuGet libraries itself (or at least something similar to this code).

304NotModified replied:
See the following examples:

These calls:

```
frameWork.Items.Add(new NuGetTargetFrameworkItem { Name = "YYYY", ShortName = "XXXXX", Framework = Frameworks.FindByIdentifierShortName("XXXXX").FirstOrDefault() });
PortableFrameworks.Add(frameWork);
```

are duplicate (code duplication)

The identifier "xamarinios" etc. is duplicate (config duplication):

```
frameWork = new NuGetTargetFrameworkItem { Name = "Xamarin For iOS", ShortName = "xamarinios", Framework = Frameworks.FindByIdentifierShortName("xamarinios").FirstOrDefault() };
```

ravensorb replied:
Again, they are not duplicates :) Each line has a small different to handle the variations in the framework definitions.

304NotModified commented on PackageExplorer/Utilities/NuGetExtensions.cs:
```diff
@@ -0,0 +1,271 @@
+			var frameworkName =
+				DefaultFrameworkMappings.Instance.IdentifierShortNames.FirstOrDefault(
+					x =>
+						string.Compare(x.Value, frameworkIdentifier, StringComparison.InvariantCultureIgnoreCase) == 0 ||
```
please replace `.Compare` with `.Equals`.

ravensorb replied:
They are not the same and have the potential for null value exceptions.  Adding a null check would be just as complicated --- unless you want me to switch to using the ?. in C# 6?

304NotModified commented on PackageExplorer/Utilities/NuGetExtensions.cs:
```diff
@@ -0,0 +1,271 @@
+	/// <summary>
+	/// Class for Extensions related to NuGetFramework objects.
+	/// </summary>
+	public static class NuGetFrameworkExtensions
```
Can you add some examples (`<example`) on this methods? I have difficulties to follow the logic here

ravensorb replied:
I'll add more comments -- these were refactored into extensions specifically because the current NuGet framework is very convoluted in how it organizes its internal data structures.

304NotModified commented on PackageExplorer/ViewModels/PortableLibraryViewModel.cs:
```diff
@@ -0,0 +1,50 @@
+
+namespace PackageExplorer.ViewModels
+{
+	public class PortableLibraryViewModel
```
Please a XML comments why this class is needed.

ravensorb replied:
This class is part of the MVVM pattern that allowed for separation of the code from the UI.  I'll add some comments to this effect.

ravensorb replied:
Done

304NotModified commented on PackageExplorer/Wpf/NuGetTargetFrameworkItemDataTemplateSelector.cs:
```diff
@@ -0,0 +1,44 @@
+	/// <summary>
+	/// Class PortableClassDataTemplateSelector.
+	/// </summary>
+	public class NuGetTargetFrameworkItemDataTemplateSelector : DataTemplateSelector
```
Please add some valuable comments.

ravensorb replied:
Done - also added a "see" comment to related it to where it being used

304NotModified commented:
Hi!

I checked the code and some parts are very difficult to follow (and review). I'm sure this isn't rocket science, so need to improve that otherwise it's too difficult to maintain this code.

ravensorb commented:
Would you like me to pull the extension methods out in to into a separate nuget package?

304NotModified commented:
What do you mean?

ravensorb commented:
the static class NuGetFrameworkExtensions -- If you have concerns about that being too complicated for the solution, I would be more than willing to create a new NuGet package (That I maintain) that contains those extensions.  Then NuGetPackageExplorer could just reference that nuget package rather than have the classes in your branch.  Just a thought :)  Otherwise, I think I updated everything else based on your feedback.

304NotModified commented:
@ravensorb 

> the static class NuGetFrameworkExtensions -- If you have concerns about that being too complicated for the solution, I would be more than willing to create a new NuGet package (That I maintain) that contains those extensions.

That's a funny idea. But isn't the case that is too complicated for me, but it's too complicated for it's purpose. I'm pretty sure you can refactor the stuff to get some improved code quality.

ravensorb commented:
You are welcome to try -- that was the best I could come up with given the nature of the Nuget framework code :smile: They have things split into multiple different collections with only a "loose" relationship (its almost like they took a normalized database and turned it into a class structure).

304NotModified commented:
> You are welcome to try

OK, but if you can fix first that it builds on AppVeyor ;) 

I think you need to enabled nuget restore for the project.

304NotModified commented:
@ravensorb can you fix the build?

ravensorb commented:
Sure -- can you give me access to the appveryor build env?

304NotModified commented:
I think it's easier that you will test your fork in appveyor (set up appveyor is done by appveyor.yml)

You sign in on appveyor (with your github account) and add you fork :) ("add project"). If you need some help on that, just let me know.

ravensorb commented:
if you add a "nuget restore" in the before build, it should work

https://www.appveyor.com/blog/2014/03/18/about-nuget-package-restore

304NotModified commented:
yes, so please add it to appveyor.yml and add in the PR :)

ravensorb commented:
Just added the support for package restore in the appveyor file

304NotModified commented:
Can you fix the merge conflict? Hopefully it builds then

304NotModified commented:
I have tested it, works good.

![image](https://cloud.githubusercontent.com/assets/5808377/19706211/ffe3f4e8-9b10-11e6-9884-70c372695bd4.png)


The merge conflict is easy to solve.

The code duplication is a todo.

304NotModified commented:
fixed the nuget issue. Ancient nuget.exe in the root (v1.5...)

304NotModified commented:
superseded by https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/pull/138
